music/src/ddsp/model.ts

```diff
@@ -0,0 +1,243 @@
+    return await memCheck();
+  }
+
+  /**
+   * Synthesizes audio features based on model.
+   *
+   * @param inputAudioBuffer An AudioBuffer
+   * @returns Audio Features of provided audio buffer
+   */
+  async synthesize(
+    audioFeatures: AudioFeatures,
+    settings?: ModelValues
+  ): Promise<Float32Array> {
+    if (settings !== null) {
+      this.settings = { ...this.settings, ...settings };
+    }
+
+    const { f0_hz, loudness_db, confidences } = audioFeatures;
+
+    // we need to upsample pitches because spice
+    // gives back something that is smaller
+    const upsampledPitches = upsample_f0(
+      f0_hz,
+      loudness_db.length,
+      this.settings.modelMaxFrameLength
+    );
+    const upsampledConfidences = upsample_f0(
+      confidences,
+      loudness_db.length,
+      this.settings.modelMaxFrameLength
+    );
+
+    const normalizedAudioFeatures = await normalizeAudioFeatures(
+      {
+        f0_hz: upsampledPitches,
+        loudness_db,
+        confidences: upsampledConfidences,
+      },
+      this.settings
+    );
+
+    // we have to process the audio in chunks according
+    // to the model max frame length
+    // eg. if we have a 10s input and a 5s model, we'll have
+    // to splice our input into 2, process it twice, and then stitch it
+    // stitching the inputs require a linear mask to prevent an audible blip
+    // we overlap the last second and the first second of 2 conjoining chunks.
+    // eg. for a 10s input [1,2,3,4,5] + [5,6,7,8,9] + [9,10]
+    const audioChunks = [];
+    const inputFrameLength = normalizedAudioFeatures.loudness_db.length;
+    const inputAudioLengthInSecs = convertFrameToSecs(inputFrameLength);
+    let isLastChunks = false;
+    const usableAudioDuration = convertFrameToSecs(
+      this.settings.modelMaxFrameLength
+    );
+    let audioBuffer;
+
+    const mixBufferLength = CROSSFADE_DURATION * MODEL_SAMPLE_RATE;
+    for (
+      let i = 0;
+      i < inputAudioLengthInSecs;
+      i += usableAudioDuration - CROSSFADE_DURATION
+    ) {
+      const startFrameIndex = Math.floor(convertSecsToFrame(i));
+      const endFrameIndex = convertSecsToFrame(i + usableAudioDuration);
+      if (endFrameIndex > inputFrameLength) {
+        isLastChunks = true;
+      }
+
+      // tslint:disable-next-line: variable-name
+      const { f0_hz, loudness_db }: AudioFeatures = await resizeAudioFeatures(
+        normalizedAudioFeatures,
+        startFrameIndex,
+        endFrameIndex,
+        isLastChunks,
+        inputFrameLength
+      );
+
+      // tslint:disable-next-line: variable-name
+      const f0hzTensor = tf.tensor1d(f0_hz, 'float32');
+      // tslint:disable-next-line: variable-name
+      const loudnessTensor = tf.tensor1d(loudness_db, 'float32');
+
+      const result = await this.model.predict({
+        f0_hz: f0hzTensor,
+        loudness_db: loudnessTensor,
+      });
+
+      // apply a loudness modifier here to make some
```

Comment:
wouldn't it make sense to do this once on the stitched audio instead of on each chunk?

Reply:
yes, good point! however, looking at the code, it seems like we're doing this mul on a tensor and then disposing it once we get the data. 

if we do it once at the end, we'll have to convert audioChunks into a tensor, run the mul, then take the data again.  not sure if there is a performance gain if we do that.

Reply:
okay, I was just looking for code clarity mostly. If you think it will look/be worse, no worries.

Reply:
word. i've made the change. I realized that it was a simple multiplier and doesnt need a tensor to do this operation.